filebeat/input/filestream/filestream.go

```diff
@@ -179,22 +179,29 @@ func (f *logFile) shouldBeClosed() bool {
 
 	info, statErr := f.file.Stat()
 	if statErr != nil {
+		// return early if the file does not exist anymore and the reader should be closed
+		if f.closeRemoved && errors.Is(statErr, os.ErrNotExist) {
+			f.log.Debugf("close.on_state_change.removed is enabled and file %s has been removed", f.file.Name())
+			return true
+		}
+
+		// If an unexpected error happens we keep the reader open hoping once everything will go back to normal.
 		f.log.Errorf("Unexpected error reading from %s; error: %s", f.file.Name(), statErr)
```

Comment:
This one still might produce a false-positive log message. The `shouldBeClosed` method is called by a separate go-routine independent from the reader go-routine. When the reader go-routine shuts down (cancelled event) while `shouldBeClosed` is run we will see an error telling us that `f` is invalid because it was already closed.

In order to prevent the false-positive message, we either need a mutex or some ref-counting on `f` + a wrapper on the reader that allows us to `unblock` a pending read call if the file gets closed async.

Reply:
Let's try to solve the race in a separate PR.